lucene/core/src/java/org/apache/lucene/search/FilteringFieldComparator.java

```diff
@@ -0,0 +1,350 @@
+        private final boolean reverse;
+        private boolean hasTopValue = false;
+        private PointValues pointValues;
+        private final int bytesCount;
+        private final byte[] minValueAsBytes;
+        private final byte[] maxValueAsBytes;
+        private boolean minValueExist = false;
+        private boolean maxValueExist = false;
+        private int maxDoc;
+        private int maxDocVisited;
+        private int updateCounter = 0;
+        private final String field;
+        protected boolean canUpdateIterator = false; // set to true when queue becomes full and hitsThreshold is reached
+
+        public FilteringNumericComparator(NumericComparator<T> in, boolean reverse, int bytesCount) {
+            super(in);
+            this.field = in.field;
+            this.bytesCount = bytesCount;
+            this.reverse = reverse;
+            minValueAsBytes = new byte[bytesCount];
+            maxValueAsBytes = new byte[bytesCount];
+            if (reverse) {
+                minValueExist = true;
+            } else {
+                maxValueExist = true;
+            }
+        }
+
+        @Override
+        public void setCanUpdateIterator() throws IOException {
+            this.canUpdateIterator = true;
+            // for the 1st time queue becomes full and hitsThreshold is reached
+            // we can start updating competitive iterator
+            updateCompetitiveIterator();
+        }
+
+        @Override
+        public void setTopValue(T value) {
+            hasTopValue = true;
+            if (reverse) {
+                maxValueExist = true;
+            } else {
+                minValueExist = true;
+            }
+            in.setTopValue(value);
+        }
+
+        @Override
+        public void setBottom(int slot) throws IOException {
+            ((NumericComparator) in).setBottom(slot);
+            updateCompetitiveIterator(); // update an iterator if we set a new bottom
+        }
+
+        @Override
+        public int compareBottom(int doc) throws IOException {
+            return ((NumericComparator) in).compareBottom(doc);
+        }
+
+        @Override
+        public int compareTop(int doc) throws IOException {
+            return ((NumericComparator) in).compareTop(doc);
+        }
+
+        @Override
+        public void copy(int slot, int doc) throws IOException {
+            ((NumericComparator) in).copy(slot, doc);
+        }
+
+        @Override
+        public void setScorer(Scorable scorer) throws IOException {}
+
+        @Override
+        public final LeafFieldComparator getLeafComparator(LeafReaderContext context) throws IOException {
+            ((NumericComparator) in).doSetNextReader(context);
+            pointValues = context.reader().getPointValues(field);
+            iterator = pointValues == null ? null : ((NumericComparator)in).currentReaderValues;
```

Comment:
I don't think we should steal the iterator of the wrapped comparator, it's a hacky and I believe that it could cause bugs due to the iterator being advanced on doc IDs out-of-order since the same iterator has two consumers that don't know about each other.

Reply:
Furthermore we use `iterator.cost() >>> 3` below as a condition. I think that the first time we should instead use the cost of the scorer otherwise this approach is going to slow down selective queries.

Reply:
addressed in  24c94ff